File: sdks/go/examples/nodes/helloworld.go
package nodes

import (
	"fmt"
	"time"

	nanoservice "github.com/nickincloud/nanoservice-go"
)

// HelloWorldNode is a simple example node that greets the user.
type HelloWorldNode struct{}

// Execute processes the hello-world node.
func (n *HelloWorldNode) Execute(ctx *nanoservice.Context, config map[string]interface{}) (interface{}, error) {
	// Get name from request body, default to "World"
	name := "World"
	if body := ctx.Request.BodyMap(); body != nil {
		if v, ok := body["name"].(string); ok && v != "" {
			name = v
		}
	}

	// Get prefix from config, default to "Hello"
	prefix := "Hello"
	if config != nil {
		if v, ok := config["prefix"].(string); ok && v != "" {
			prefix = v
		}
	}

	message := fmt.Sprintf("%s, %s!", prefix, name)

	// Store greeting in context vars for downstream nodes
	ctx.SetVar("greeting", message)

	return map[string]interface{}{
		"message":   message,
		"timestamp": time.Now().UTC().Format(time.RFC3339),
		"language":  "go",
	}, nil
}

File: sdks/rust/src/nodes/hello_world.rs
use async_trait::async_trait;
use std::collections::HashMap;

use crate::node::NodeHandler;
use crate::types::Context;

/// HelloWorldNode greets the user with a configurable prefix.
pub struct HelloWorldNode;

#[async_trait]
impl NodeHandler for HelloWorldNode {
    async fn execute(
        &self,
        ctx: &mut Context,
        config: &HashMap<String, serde_json::Value>,
    ) -> Result<serde_json::Value, Box<dyn std::error::Error + Send + Sync>> {
        let name = ctx
            .request
            .body
            .get("name")
            .and_then(|v| v.as_str())
            .unwrap_or("World");

        let prefix = config
            .get("prefix")
            .and_then(|v| v.as_str())
            .unwrap_or("Hello");

        let message = format!("{}, {}!", prefix, name);

        ctx.set_var("greeting", serde_json::Value::String(message.clone()));

        Ok(serde_json::json!({
            "message": message,
            "timestamp": chrono::Utc::now().to_rfc3339(),
            "language": "rust"
        }))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testing::MockContext;

    #[tokio::test]
    async fn test_hello_world_default() {
        let node = HelloWorldNode;
        let mut ctx = MockContext::new().build();
        let result = node.execute(&mut ctx, &HashMap::new()).await.unwrap();
        assert_eq!(result["message"], "Hello, World!");
        assert_eq!(result["language"], "rust");
    }

    #[tokio::test]
    async fn test_hello_world_with_name() {
        let node = HelloWorldNode;
        let mut ctx = MockContext::new()
            .with_body(serde_json::json!({"name": "Blok"}))
            .build();
        let result = node.execute(&mut ctx, &HashMap::new()).await.unwrap();
        assert_eq!(result["message"], "Hello, Blok!");
    }

    #[tokio::test]
    async fn test_hello_world_with_prefix() {
        let node = HelloWorldNode;
        let mut ctx = MockContext::new()
            .with_body(serde_json::json!({"name": "World"}))
            .build();
        let config = HashMap::from([("prefix".into(), serde_json::json!("Hi"))]);
        let result = node.execute(&mut ctx, &config).await.unwrap();
        assert_eq!(result["message"], "Hi, World!");
    }

    #[tokio::test]
    async fn test_hello_world_sets_var() {
        let node = HelloWorldNode;
        let mut ctx = MockContext::new().build();
        node.execute(&mut ctx, &HashMap::new()).await.unwrap();
        assert!(ctx.get_var_str("greeting").is_some());
    }
}
